src/content/en/2021/webassembly.md

```diff
@@ -15,6 +16,461 @@ featured_stat_2: TODO
+  description="Bar chart showing `wasm-opt` size benchmarks across a wide variety of real-world modules such as `base64`, `box2d`, `lua`, `sqlite`, `zlib`, etc. The resulting sizes vary from 78% to, worst-case, 93% of the original. Median is at 89%.",
+  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vT6yhkn3lw148YQQHLoqA71NIsZLSSoBtgFmd_hRyhcmyPl2OpLyuOjUBk64I5DLE_grN8esL8oA3zt/pubchart?oid=724059342&format=interactive",
+  sheets_gid="1763590037",
+  image="wasm_opt_bench.png",
+  width=600,
+  height=615
+  )
+}}
+
+{{ figure_markup(
+  caption="`wasm-opt` + Brotli size benchmarks.",
+  description="Another bar chart showing same `wasm-opt` size benchmarks, but with Brotli compression applied to both original and resulting WebAssembly modules. Results vary from 83% to 99% with median at 91%.",
+  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vT6yhkn3lw148YQQHLoqA71NIsZLSSoBtgFmd_hRyhcmyPl2OpLyuOjUBk64I5DLE_grN8esL8oA3zt/pubchart?oid=2047888306&format=interactive",
+  sheets_gid="1763590037",
+  image="wasm_opt_br_bench.png",
+  width=600,
+  height=615
+  )
+}}
+
+We've decided to check the performance of `wasm-opt` on the collected HTTP Archive dataset as well, but there's a catch.
+
+As mentioned above, `wasm-opt` is already used by most compiler toolchains, so most of the modules in the dataset are already its resulting artifacts. Unlike in compression analysis above, there's no way for us to reverse existing optimizations and run `wasm-opt` on the originals. Instead, we're re-running `wasm-opt` on pre-optimized binaries, which skews the results. This is the command we've used on binaries produced after the strip-debug step:
+
+```bash
+wasm-opt -O -all some.wasm -o some.opt.wasm
+```
+
+Then, we compressed the results to Brotli and compared to the previous step, as usual.
+
+While the resulting data is not representative of real-world usage and not relevant to regular consumers who should use `wasm-opt` as they normally do, it might be useful to consumers like CDNs that want to run optimizations at scale, as well as to the Binaryen team itself:
+
+{{ figure_markup(
+  caption="`wasm-opt` + Brotli savings.",
+  description="Bar chart showing the distribution of absolute size changes when `wasm-opt` + Brotli are executed against our modified datasets. The results are mixed—0 percentile shows a 260 KB regression on desktop and 180 KB on mobile, 10% shows 27 KB and 6 KB regressions, median is at 3.6 KB / 0.6 KB regressions, 90% shows 513 bytes improvement on desktop and 77 bytes improvement on mobile, and maximum savings end up at 1 MB on desktop and almost 250 KB on mobile.",
+  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vT6yhkn3lw148YQQHLoqA71NIsZLSSoBtgFmd_hRyhcmyPl2OpLyuOjUBk64I5DLE_grN8esL8oA3zt/pubchart?oid=541095203&format=interactive",
+  sheets_gid="1401232418",
+  sql_file="sizes_and_savings.sql",
+  image="wasm_opt_br_savings.png"
+  )
+}}
+
```

Comment:
Similar comment to above. Though think you've explained it better in the text here.

![image](https://user-images.githubusercontent.com/10931297/142759922-7fa3c220-ce81-4a6d-9388-0fda94d73e34.png)

percentile | desktop | mobile
-- | -- | --
0 | -260,632 | -178,826
1 | -118,704 | -66,562
2 | -118,704 | -49,206
3 | -118,704 | -39,136
4 | -118,704 | -27,119
5 | -85,564 | -20,875
6 | -50,169 | -16,809
7 | -43,979 | -16,236
8 | -29,499 | -10,592
9 | -27,119 | -5,987
10 | -27,119 | -5,890
11 | -25,982 | -5,890
12 | -16,809 | -5,890
13 | -14,181 | -5,890
14 | -9,194 | -5,890
15 | -5,890 | -5,890
16 | -5,890 | -5,890
17 | -5,890 | -5,890
18 | -5,890 | -5,890
19 | -5,890 | -5,890
20 | -5,890 | -5,890
21 | -5,890 | -5,890
22 | -5,890 | -5,890
23 | -5,890 | -5,890
24 | -5,890 | -5,890
25 | -5,890 | -5,775
26 | -5,890 | -5,775
27 | -5,890 | -5,775
28 | -5,890 | -5,775
29 | -5,890 | -5,775
30 | -5,890 | -5,775
31 | -5,890 | -5,775
32 | -5,890 | -5,775
33 | -5,890 | -5,775
34 | -5,890 | -5,775
35 | -5,890 | -5,775
36 | -5,890 | -5,775
37 | -5,890 | -5,656
38 | -5,775 | -4,850
39 | -5,775 | -4,788
40 | -5,775 | -4,788
41 | -5,775 | -4,305
42 | -5,775 | -4,142
43 | -5,775 | -3,647
44 | -5,775 | -2,792
45 | -4,802 | -2,419
46 | -4,788 | -2,419
47 | -4,456 | -2,306
48 | -4,207 | -1,754
49 | -3,877 | -1,117
50 | -3,647 | -611
51 | -2,792 | -492
52 | -2,419 | -268
53 | -2,306 | -220
54 | -1,945 | -119
55 | -1,167 | -119
56 | -1,117 | -38
57 | -569 | -12
58 | -492 | -12
59 | -248 | -12
60 | -119 | -12
61 | -119 | -12
62 | -45 | -12
63 | -12 | -12
64 | -12 | -11
65 | -12 | -7
66 | -12 | -7
67 | -12 | -1
68 | -12 | 0
69 | -10 | 0
70 | -7 | 0
71 | 0 | 0
72 | 0 | 2
73 | 0 | 3
74 | 0 | 5
75 | 2 | 6
76 | 4 | 7
77 | 5 | 10
78 | 10 | 16
79 | 16 | 22
80 | 22 | 22
81 | 28 | 30
82 | 30 | 30
83 | 32 | 32
84 | 32 | 32
85 | 32 | 32
86 | 34 | 32
87 | 77 | 32
88 | 187 | 32
89 | 314 | 34
90 | 513 | 77
91 | 730 | 246
92 | 1,440 | 443
93 | 2,626 | 600
94 | 5,556 | 1,039
95 | 7,171 | 3,135
96 | 11,910 | 5,556
97 | 12,938 | 7,171
98 | 24,465 | 11,316
99 | 24,465 | 12,938
100 | 1,032,341 | 246,911

Reply:
I think this one can be converted into 10-90 range - it should look fine without log scale - and then I'll mention the extremes in text. Can you convert the graph into the same style as you did others please?

Reply:
Done, committed and staging update: https://20211121t105115-dot-webalmanac.uk.r.appspot.com/en/2021/webassembly#fig-18

Though not showing much when converted to kilobytes:

![image](https://user-images.githubusercontent.com/10931297/143062207-4c09abf5-bfe0-49fb-80e5-313115c10a7d.png)

Should we revert to bytes for this one? Though that feels a little bit of a con just to inflate the numbers... 😀

![image](https://user-images.githubusercontent.com/10931297/143062423-3377292a-1aa6-49af-96a4-a2bea5b0e621.png)

Reply:
I guess we could keep KB but just add 2 (?) digits after a dot as a compromise between consistency (KB in all graphs) and usefulness.

Reply:
Also see my comment above - should we maybe just revert to bytes for those? It makes it clearer when e.g. zero is actually zero and not a rounding error.

Reply:
Good idea! Also had my rang wrong so missed of the 10th percentile and messed up my desktop and mobile labels:

![image](https://user-images.githubusercontent.com/10931297/143066604-f7f4bc03-a87d-455e-89f2-bceed6da7d2c.png)

Reply:
On this one I do think converting to bytes makes it more difficult to read:

![image](https://user-images.githubusercontent.com/10931297/143067615-24605611-92d6-45fc-8989-3634469af645.png)

But happy to leave it in bytes if you want?

Reply:
Hmm yeah overlapping labels are definitely not great. I guess we could change font size, but that only makes things worse.

The 2 digits after dot looks good enough for a compromise, let's leave that.

Reply:
Marking this one as resolved as it's getting hard to track conversations 😅